docs/docs/typescript/utility-types.md

```diff
@@ -0,0 +1,222 @@
+
+Not only does `CellSuccessProps` type the data returned from the query, but it also types the variables and methods returned by Apollo Client's `useQuery` hook!
+
+```ts title="web/src/components/BlogPost.cell.tsx"
+import type { FindBlogPostQuery, FindBlogPostQueryVariables } from 'types/graphql'
+
+// highlight-next-line
+import type { CellSuccessProps } from '@redwoodjs/web'
+
+// ...
+
+// highlight-next-line
+type SuccessProps = CellSuccessProps<FindBlogPostQuery, FindBlogPostQueryVariables>
+
+export const Success = ({
+  blogPost, // From the query. This is typed of course
+  refetch, // 👈 From Apollo Client. This is typed too!
+  fetchMore
+// highlight-next-line
+}: SuccessProps) => {
+  // ...
+}
+```
+
+### `CellFailureProps<TVariables>`
+
+This gives you the types of the props in your Cell's `Failure` component.
+It takes `TVariables` as an optional generic parameter, which is useful if you want to print error messages like `"Couldn't load data for ${variables.searchTerm}"`:
+
+```ts title=web/src/components/BlogPost.cell.tsx
+import type { FindBlogPostQuery, FindBlogPostQueryVariables } from 'types/graphql'
+
+// highlight-next-line
+import type { CellFailureProps } from '@redwoodjs/web'
+
+// ...
+
+export const Failure = ({
+  error,
+  variables // 👈 Variables is typed based on the generic
+  // highlight-next-line
+}: CellFailureProps<FindBlogPostQueryVariables>) => (
+  // ...
+)
+```
+
+### `CellLoadingProps<TVariables>`
+
+Similar to `CellFailureProps`, but for the props of your Cell's `Loading` component:
+
+```ts title=web/src/components/BlogPost.cell.tsx
+import type { FindBlogPostQuery, FindBlogPostQueryVariables } from 'types/graphql'
+
+// highlight-next-line
+import type { CellLoadingProps } from '@redwoodjs/web'
+
+// ...
+
+// highlight-next-line
+export const Loading = (props: CellLoadingProps<FindBlogPostQueryVariables>) => (
+  <div>Loading...</div>
+)
+```
+
+## Scenarios & Testing
+
+Over on the api side, when you generate SDLs and Services, Redwood generates tests and scenarios with all the types required. Let's take a deeper look at scenario types.
+
+### `defineScenario`
+
+This is actually a function, not a type, but it takes a lot of generics. Use as many or as little as you find helpful.
+
+```
+defineScenario<PrismaCreateType, TName, TKey>
+```
+
+| Generic            | Description                                                                                         |
+| :----------------- | :-------------------------------------------------------------------------------------------------- |
+| `PrismaCreateType` | (Optional) the type imported from Prisma's create operation that goes into the "data" key           |
+| `TName`            | (Optional) the name or names of the models in your scenario                                         |
+| `TKey`             | (Optional) the keys in your scenario. These are really only useful while you write out the scenario |
```

Comment:
There's some confusion between `TKey` vs `TKeys` for `defineScenario` and `ScenarioData` and I wasn't sure which was supposed to which.

Reply:
It looks like `ScenarioData` takes `TKeys`, but the name of the generics in the src for `defineScenario` are a bit differently-named from what was written up here:

```ts
    PrismaCreateType extends { data: any },
    ModelName extends string | number | symbol = string | number | symbol,
    KeyName extends string | number | symbol = string | number | symbol
```

Reply:
Updated so that the generic names are the same!